src/main/java/com/redhat/pantheon/servlet/DefaultQueryServlet.java

```diff
@@ -0,0 +1,107 @@
+
+import static com.redhat.pantheon.servlet.ServletUtils.paramValue;
+import static com.redhat.pantheon.servlet.ServletUtils.paramValueAsLong;
+import static com.redhat.pantheon.servlet.ServletUtils.writeAsJson;
+
+/**
+ * This query servlet applies to any resource in the content tree. It enables an action under the resource
+ * to search for nodes. It will search for all nodes under the requested resource, not just direct descendants.
+ * Currently the only parameter being used is the node name.
+ * <p>
+ * This servlet is enabled by default for the 'query' selector and 'json' extension.
+ * </p>
+ * <p>
+ * Child nodes can now be searched at the following endpoint (assuming the node location is /content/mynode):
+ * </p>
+ * <p>
+ * GET /content/mynode.query.json?...
+ * <p>
+ * </p>
+ * Available query parameters for this servlet are:
+ * <br>
+ * nodeType (optional): The JCR node type to look for. Default value: 'nt:base'
+ * <br>
+ * where (optional): Any additional JCR-QL2 where statements to further filter the results.
+ * e.g. '[app:my-property] = "prop value"'. Default value: '' (no additional filtering will be performed)
+ * <br>
+ * orderBy (optional): A JCR-QL2 order by statement to sort the results. e.g. '[app:my-property] desc'.
+ * Default value: '' (no sorting will be performed)
+ * <br>
+ * limit (optional): The maximum number of results to return. Default value: 1000
+ * <br>
+ * offset (optional): The number of results to skip (for pagination). Default value: 0
+ * <p>
+ * </p>
+ */
+@Component(
+        service = Servlet.class,
+        property = {
+                "sling.servlet.resourceTypes=sling/servlet/default",
+                "sling.servlet.selectors=query",
+                "sling.servlet.extensions=json",
+                Constants.SERVICE_DESCRIPTION + "=Utility servlet capable of running JCR queries",
+                Constants.SERVICE_VENDOR + "=Red Hat Content Tooling team"
+        })
+public class DefaultQueryServlet extends SlingSafeMethodsServlet {
+    private final Logger logger = LoggerFactory.getLogger(DefaultQueryServlet.class);
+
+    private static final long RESULT_SIZE_LIMIT = 1000;
+
+    @Override
+    protected void doGet(@Nonnull SlingHttpServletRequest request, @Nonnull SlingHttpServletResponse response)
+            throws ServletException, IOException {
+
+        // Get the query parameter(s)
+        String nodeType = paramValue(request, "nodeType","nt:base");
+        String where = paramValue(request, "where");
+        String orderBy = paramValue(request, "orderBy");
+        long limit = paramValueAsLong(request, "limit", RESULT_SIZE_LIMIT);
+        long offset = paramValueAsLong(request, "offset", 0L);
+
+        StringBuilder query = new StringBuilder("select * from ")
+                .append("[").append(nodeType).append("]")
+                .append(" as node ")
+                .append(" where ISDESCENDANTNODE(\"" + request.getResource().getPath() + "\") ");
```

Comment:
How is the user prevented from querying for nodes that are outside of the /content path? It looks like this might list all child nodes for any resource in the system, not just those that are in /content. If we're relying on Sling ACLs for that, then we should be diligent about knowing exactly what nodes users are able to query for using this tool.

(On that note, it's probably worth it to add some negative unit tests as well - to ensure that users can't get a listing of nodes that we know to be sensitive - underneath /home/users for example...)

Reply:
The access to the resource itself is in fact controlled by the sling ACLs. So only users with read access to the node will be able to use this servlet. Additionally, the `ISDESCENDANTNODE` clause (which is hard coded) restricts all returned results to nodes under the requested one. So, in your example, because `/home/users` is a restricted node (by ALCs) no user would be able to get to this servlet.

The only problem I foresee is if for some reason the child nodes are restricted (as in not even viewable) by the user, but not the parent node, and hence said restricted nodes could be returned by this servlet.

Reply:
I would imagine that the query is still running in the context of the user, so it wouldn't return any restricted child nodes.

I'm still uneasy about this because I don't know if we're opening up a security concern, but it's mostly because I'm unfamiliar with 1.) how sling's acls are set up ootb, and 2.) what data even exists that might be considered sensitive. I'll ignore these concerns for now - my favorite phrase "don't solve problems you don't have yet" comes to mind.